accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -826,9 +829,18 @@ _is the extension type_
 <code>V\<T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
 and that its static type _is an extension type_.
 
+We say that an extension type is _unrelated to `Future`_ in the case where
+it does not implement `Future<U>` for any `U` *(this means that it has no
+direct or indirect superinterface of the form `Future<...>`)*.
+
 It is a compile-time error if `await e` occurs, and the static type of
-`e` is an extension type which is not a subtype of `Future<T>` for any
-`T`.
+`e` satisfiers at least one of the following criteria:
+
+- an extension type which is unrelated to `Future`.
+- a type of the form `T?` or `FutureOr<T>` where `T` is a type that matches
```

Comment:
We think we should not include `FutureOr<T>` here. The type `FutureOr<T>` is very much releated to `Future`, even if `T` isn't. Actually, *especially* if `T` isnt.

You should be allowed to await `FutureOr<Ext>` where `Ext` is an extension type unrelated to `Future`. 
We want to allow you to await `FutureOr<T>` for any `T`, because otherwise you'd have to do an `(v is Future<T>) ? await v : v`, which would be allowed and have all the same issues as what we are trying to avoid by disallowing it.

Reply:
Well, the point was that if you have an expression of type `E` which is an extension type that has no relation to `Future` then you can trivially type that same expression as `FutureOr<E>`, which we can think of as `Future<E> | E, and then it seems contradictory if it's ok to await that, but it is not OK to await `E` alone. So I classified all the unions including the "don't await me" type as "don't await me" as well.

We could drop this whole idea and say that (1) await just uses the extension type erasure, no matter what (also for computation of the type of `await e`), and (2) it's up to a lint to tell people that it's a violation of some notion of encapsulation to have this particular await expression.